lang/ir/src/ir/item_impl/callable.rs

```diff
@@ -0,0 +1,557 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utilities and helper routines that are useful for both ink! messages
+//! and ink! constructors.
+
+use crate::ir;
+use core::fmt;
+use proc_macro2::Ident;
+use quote::ToTokens as _;
+
+/// The kind of externally callable smart contract entity.
+pub(super) enum CallableKind {
+    /// An ink! message externally callable.
+    Message,
+    /// An ink! constructor externally callable.
+    Constructor,
+}
+
+impl fmt::Display for CallableKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Message => write!(f, "message"),
+            Self::Constructor => write!(f, "constructor"),
+        }
+    }
+}
+
+/// Wrapper for a callable that adds its composed selector.
+pub struct CallableWithSelector<'a, C> {
+    /// The composed selector computed by the associated implementation block
+    /// and the given callable.
+    composed_selector: ir::Selector,
+    /// The actual callable.
+    callable: &'a C,
+}
+
+impl<'a, C> CallableWithSelector<'a, C>
+where
+    C: Callable,
+{
+    /// Creates a new wrapper around the given callable and parent impl block.
+    pub(super) fn new(item_impl: &ir::ItemImpl, callable: &'a C) -> Self {
+        Self {
+            composed_selector: compose_selector(item_impl, callable),
+            callable,
+        }
+    }
+}
+
+impl<'a, C> CallableWithSelector<'a, C> {
+    /// Returns the composed selector of the ink! callable the the impl block.
+    pub fn composed_selector(&self) -> ir::Selector {
+        self.composed_selector
+    }
+
+    /// Returns a shared reference to the underlying callable.
+    pub fn item(&self) -> &'a C {
+        self.callable
+    }
+}
+
+impl<'a, C> Callable for CallableWithSelector<'a, C>
+where
+    C: Callable,
+{
+    fn ident(&self) -> &Ident {
+        <C as Callable>::ident(&self.callable)
+    }
+
+    fn user_provided_selector(&self) -> Option<&ir::Selector> {
+        <C as Callable>::user_provided_selector(&self.callable)
+    }
+
+    fn is_payable(&self) -> bool {
+        <C as Callable>::is_payable(&self.callable)
+    }
+
+    fn visibility(&self) -> Visibility {
+        <C as Callable>::visibility(&self.callable)
+    }
+
+    fn inputs(&self) -> InputsIter {
+        <C as Callable>::inputs(&self.callable)
+    }
+
+    fn statements(&self) -> &[syn::Stmt] {
+        <C as Callable>::statements(&self.callable)
+    }
+}
+
+impl<'a, C> ::core::ops::Deref for CallableWithSelector<'a, C> {
+    type Target = C;
+
+    fn deref(&self) -> &Self::Target {
+        &self.callable
+    }
+}
+
+/// An ink! callable.
+///
+/// This is either an ink! message or an ink! constructor.
+/// Used to share common behavior between different callable types.
+pub trait Callable {
```

Comment:
I chose this design to avoid further nesting of items. Technically you are definitely right that we could merge those types. However, I think the current design through the trait definition is fine. Maybe we can explore this other design in some follow-up PR?

Reply:
Good enough for me - was just curious about the reasoning because I always go back and forth on this and usually *slightly* favour not using traits.